src/storage/txn/tracker.rs

```diff
@@ -0,0 +1,293 @@
+            // Prevent double counting.
+            state.current_stage = PollState::None;
+        });
+    }
+}
+
+impl State {
+    fn collect_to_tracker(&mut self, now: Instant, tracker: &mut tracker::Tracker) {
+        if let PollState::Begin(at) = self.current_stage {
+            self.future_process_nanos += now.saturating_duration_since(at).as_nanos() as u64;
+        }
+        tracker.metrics.future_process_nanos += self.future_process_nanos;
+        tracker.metrics.future_suspend_nanos += self.future_suspend_nanos;
+        self.future_process_nanos = 0;
+        self.future_suspend_nanos = 0;
+    }
+
+    fn on_poll_begin(&mut self, now: Instant) {
+        match self.current_stage {
+            PollState::None => {}
+            PollState::Finish(at) => {
+                self.future_suspend_nanos += now.saturating_duration_since(at).as_nanos() as u64;
+            }
+            _ => {
+                panic!(
+                    "unexpected tracker state {:?}, tag: {:?}, cid: {:?}, token: {:?}",
+                    self.current_stage, self.tag, self.cid, self.token,
+                )
+            }
+        }
+    }
+
+    fn on_poll_finish(&mut self, now: Instant, tracker: &mut tracker::Tracker) {
+        match self.current_stage {
+            PollState::None => {}
+            PollState::Begin(_) => {
+                self.collect_to_tracker(now, tracker);
+            }
+            _ => {
+                panic!(
+                    "unexpected tracker state {:?}, tag: {:?}, cid: {:?}, token: {:?}",
+                    self.current_stage, self.tag, self.cid, self.token,
+                );
+            }
+        }
+    }
+}
+
+impl FutureTrack for TlsFutureTracker {
+    fn on_poll_begin(&mut self) {
+        let mut state = self.state.take().expect("future tracker state is not set");
+        let now = Instant::now();
+        state.on_poll_begin(now);
+        CURRENT_STATE.with(|current_state| {
+            assert!(
+                current_state.borrow_mut().is_none(),
+                "TLS future tracker state is already set, {:?} vs {:?}",
+                current_state.borrow().as_ref().unwrap(),
+                state
+            );
+            state.current_stage = PollState::Begin(now);
+            *current_state.borrow_mut() = Some(state);
+        });
+    }
+
+    fn on_poll_finish(&mut self) {
+        CURRENT_STATE.with(|current_state| {
+            let mut state = current_state
+                .borrow_mut()
+                .take()
+                .expect("TLS future tracker state is not set");
+            let now = Instant::now();
+            GLOBAL_TRACKERS.with_tracker(state.token, |tracker| {
+                state.on_poll_finish(now, tracker);
+            });
+            // Set to Finish state to tracking the wait time before the next poll.
+            state.current_stage = PollState::Finish(now);
+            self.state = Some(state);
+        });
+    }
+}
+
+#[cfg(test)]
```

Comment:
How about adding integration test cases in `kv.rs` and verify the recorded duration is not zero?